src/afl-fuzz-queue.c

```diff
@@ -648,53 +650,125 @@ void cull_queue(afl_state_t *afl) {
+
+          }
 
         }
 
-      }
+        if (!afl->top_rated[i]->favored) {
 
-      if (!afl->top_rated[i]->favored) {
+          afl->top_rated[i]->favored = 1;
+          ++afl->queued_favored;
 
-        afl->top_rated[i]->favored = 1;
-        ++afl->queued_favored;
+          if (afl->top_rated[i]->fuzz_level == 0 ||
+              !afl->top_rated[i]->was_fuzzed) {
 
-        if (afl->top_rated[i]->fuzz_level == 0 ||
-            !afl->top_rated[i]->was_fuzzed) {
+            ++afl->pending_favored;
 
-          ++afl->pending_favored;
+          }
 
         }
 
       }
 
     }
+  } else {
+    // otherwise, randomly assign favorites
+    int r;
+    int rid;
+    double weight;
+
+    for (i = 0; i < afl->queued_paths; i++) {
+      weight = 1.0;
+      if (!afl->enable_uniformly_random_favorites) {
+        // enable_boost_inputs
+        double base_weight_fac_boost_inputs = 1.0;
+        double max_weight_fac_incr = 7.0;
+        double scale_fac_boost_inputs = 0.001;
+        double num_selections = (double)afl->queue_buf[i]->num_fuzzed;
+        weight *= base_weight_fac_boost_inputs + max_weight_fac_incr / (scale_fac_boost_inputs * num_selections + 1.0);
+
+        // enable_boost_fast_seqs
+        double base_weight_fac_boost_fast = 8.0;
+        double max_weight_fac_decr = 7.0;
+        double scale_fac_boost_fast = 0.001; // based on the past experiment, 0.001 seems to be the best candidate among 0.005, 0.0025, 0.0005
+        double execs_per_sec = 1000000.0 / (double)afl->queue_buf[i]->exec_us;
+        weight *= base_weight_fac_boost_fast - max_weight_fac_decr / (scale_fac_boost_fast*execs_per_sec + 1.0);
+      }
+      r = 0;
+      rid = INT_MAX;
+      while (weight >= 1.0) {
+        r = rand_below(afl, INT_MAX);
+        if (r < rid)
+          rid = r;
+        weight -= 1.0;
+      }
+      if (weight > 0.0 && weight > rand_double(afl)) {
+        r = rand_below(afl, INT_MAX);
+        if (r < rid)
+          rid = r;
+      }
+
+      afl->queue_buf[i]->favored = 0;
+      afl->queue_buf[i]->rand = rid;
+    }
 
+    // reset findings from the previous cycle
+    memset(afl->edge_to_minimum_entry, 0, afl->fsrv.map_size * sizeof(struct queue_entry));
+
+    // going through all entries, iteratvely check for covered edges and compare against the corresponding minimum entry
+    for (i = 0; i < afl->queued_paths; i++) {
+      if (afl->queue_buf[i]->trace_mini) {
+        for (j = 0; j < afl->fsrv.map_size; j++) {
+          if (afl->queue_buf[i]->trace_mini[j >> 3] & (1 <<(j & 7))) {
+            if (!afl->edge_to_minimum_entry[j] || afl->queue_buf[i]->rand < afl->edge_to_minimum_entry[j]->rand) {
+              afl->edge_to_minimum_entry[j] = afl->queue_buf[i];
+            }
+          }
+        }
+      }
+    }
+
+    for (i = 0; i < afl->fsrv.map_size; i++) {
+      if (afl->edge_to_minimum_entry[i] && !afl->edge_to_minimum_entry[i]->favored) {
```

Comment:
Maybe we should store `afl->edge_to_minimum_entry[i]` in a local variable to avoid multiple lookups?